File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int pre_idx = 0;
    std::unordered_map<int,int> indices;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for(int i = 0; i < inorder.size(); i++){
            indices[inorder[i]] = i;
        }

        return dfs(0,inorder.size() - 1,preorder);
    }

    TreeNode* dfs(int left, int right, const std::vector<int>& preorder){
        if(left > right){
            return nullptr;
        }

        int root_val = preorder[pre_idx++];
        TreeNode* root = new TreeNode(root_val);

        int mid = indices[root_val];

        root->left = dfs(left, mid - 1, preorder);
        root->right = dfs(mid + 1, right, preorder);
        return root;
    }
};

File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        left = 0
        right = n - 1
        mid = (right + left) // 2
        while left <= right:
            if nums[mid] > target:
                right = mid - 1
                mid = (right + left) // 2
            elif nums[mid] < target:
                left = mid + 1
                mid = (right + left) // 2
            else:
                return mid
        return -1
